### ISSUE TITLE
Numerical instability in (LSTM) training

### ISSUE THREAD

**Issue by DavidNemeskey** [pytorch/pytorch#5306]
I was trying to reproduce the numbers in the language modeling paper Zaremba et al. (2014), when I ran into some problems. I re-implemented LSTM by hand in both Pytorch and TensorFlow, and while the TF version converged, the PT version did not. I have created a [repository](https://github.com/DavidNemeskey/debug_pytorch_lm) to document my findings; please read the [readme](https://github.com/DavidNemeskey/debug_pytorch_lm/blob/master/README.md) there for all the details, as well as for information on how to reproduce the problem.

In a nutshell, the signs point to a _numerically instability_ somewhere in PT. With the learning rate(s) of the TF version, losses quickly become huge, and the curve of the learning rate -> perplexity graph is heavily serrated. I don't think that the fault lies with the LSTM, but it is possible that the persistent state exacerbates the problem.

The issue can be reproduced on both CPU and GPU, with both float and double tensors.

When submitting a bug report, please include the following information (where relevant):
- OS: Linux XXX 3.16.0-4-amd64 #1 SMP Debian 3.16.36-1+deb8u2 (2016-10-19) x86_64 GNU/Linux
- PyTorch version: 0.4.0a0 (but 3 as well)
- How you installed PyTorch (conda, pip, source): source
- Python version: 3.6.4
- CUDA/cuDNN version: 8 / 6
- GPU models and configuration: GeForce GTX 980
- GCC version (if compiling from source): Debian 4.9.2-10+deb8u1

**Comment by apaszke** [comment]
Are you sure weight initialization used in both scripts is the same?

**Comment by DavidNemeskey** [comment]
Yes, but if you read the readme, you can see that I save a set of weights first, and then load it into both models. Also, you can have a look at the "trace" outputs in the logs/ directory; though not all of the arrays are shown (default numpy clipping), you can get a taste of how the numbers change.

**Comment by soumith** [comment]
The way you presented this is SUPER helpful. We'll look into this with a high-priority and figure out what's going on.

**Comment by ezyang** [comment]
Hi @DavidNemeskey, I have a few questions about the script you uploaded.

1. The TensorFlow script does not appear to be deterministic.

```
RUN 1:
Loaded parameters from params.npz
| epoch   1 |   200/ 2323 batches | lr 1.00 | ms/batch 27.34 | loss  6.89 | ppl   984.01
| epoch   1 |   400/ 2323 batches | lr 1.00 | ms/batch 23.61 | loss  6.26 | ppl   522.35
| epoch   1 |   600/ 2323 batches | lr 1.00 | ms/batch 34.92 | loss  5.94 | ppl   379.29

RUN 2:
Loaded parameters from params.npz
| epoch   1 |   200/ 2323 batches | lr 1.00 | ms/batch 28.18 | loss  6.89 | ppl   983.19
| epoch   1 |   400/ 2323 batches | lr 1.00 | ms/batch 23.94 | loss  6.27 | ppl   527.35
```

Do you know what is causing this nondeterminism? This is on TensorFlow 1.4 with GPU support. The PyTorch script, though divergent, is deterministic.

```
RUN 1:
Loaded parameters from params.npz
| epoch   1 |   200/ 2323 batches | lr 1.00 | ms/batch 105.15 | loss 19.87 | ppl 427251983.10
| epoch   1 |   400/ 2323 batches | lr 1.00 | ms/batch 97.40 | loss 19.27 | ppl 233238198.57
| epoch   1 |   600/ 2323 batches | lr 1.00 | ms/batch 99.69 | loss 22.62 | ppl 6660614197.29
| epoch   1 |   800/ 2323 batches | lr 1.00 | ms/batch 99.81 | loss 22.40 | ppl 5335789184.99
| epoch   1 |  1000/ 2323 batches | lr 1.00 | ms/batch 106.92 | loss 22.22 | ppl 4445882129.82
| epoch   1 |  1200/ 2323 batches | lr 1.00 | ms/batch 114.48 | loss 20.94 | ppl 1243664530.65

RUN 2:
Loaded parameters from params.npz
| epoch   1 |   200/ 2323 batches | lr 1.00 | ms/batch 111.11 | loss 19.87 | ppl 427251983.10
| epoch   1 |   400/ 2323 batches | lr 1.00 | ms/batch 111.10 | loss 19.27 | ppl 233238198.57
| epoch   1 |   600/ 2323 batches | lr 1.00 | ms/batch 105.96 | loss 22.62 | ppl 6660614197.29
| epoch   1 |   800/ 2323 batches | lr 1.00 | ms/batch 108.41 | loss 22.40 | ppl 5335789184.99
| epoch   1 |  1000/ 2323 batches | lr 1.00 | ms/batch 108.93 | loss 22.22 | ppl 4445882129.82
```

2. As you can see above, when I run the PyTorch script, I get perplexities that are far greater than those recorded in your experiments. This result seems to be robust even when I reinitialize the parameters (by omitting the `-L` parameter and passing a different `--seed`).

```
(/home/ezyang/Dev/pytorch-env) [ezyang@devgpu005.ash6 ~/Dev/oss/debug_pytorch_lm] python scripts/zaremba_pytorch.py -c -d ptb --seed 0
| epoch   1 |   200/ 2323 batches | lr 1.00 | ms/batch 111.69 | loss 206.33 | ppl 405118226085607054480875801125402120406150439251323597652139845617320125818060519996653568.00
(/home/ezyang/Dev/pytorch-env) [ezyang@devgpu005.ash6 ~/Dev/oss/debug_pytorch_lm] python scripts/zaremba_pytorch.py -c -d ptb --seed 1
| epoch   1 |   200/ 2323 batches | lr 1.00 | ms/batch 110.40 | loss 45.12 | ppl 39377511868908847104.00
```

Did you see this when running PyTorch? FWIW my commit hash is roughly equivalent to 031412a14b7c20e9030b00c26f30c5524a0eb028

3. In your TF script, you have written LSTM in the following way:

```
        i = tf.sigmoid(ifgo[:, :self.hidden_size])
        f = tf.sigmoid(ifgo[:, self.hidden_size:2*self.hidden_size])
        g = tf.tanh(ifgo[:, 2*self.hidden_size:3*self.hidden_size])
        o = tf.sigmoid(ifgo[:, 3*self.hidden_size:])
```

Is there any particular reason you did it this way instead of using `tf.split`?

Thanks!

**Comment by DavidNemeskey** [comment]
@ezyang I'll try to answer them. :)

1. I have no idea why this is the case. I am wondering if it has to do with the random seed not working somehow (or maybe the code also needs `np.random.seed`) -- even though I load the parameters from file, so there shouldn't be anything random left in it... But I can confirm that the phenomenon exists. Maybe ask on the TensorFlow forums (do they have one?).

2. The three PPLs for the three seeds that you show in your comment are very different from one another, so I wouldn't wonder if the difference between our numbers came down to the GPU and library versions used (CUDA, cuDNN, pytorch, g++, etc.)

3. Senility, I guess. But I do accept pull requests. :)

**Comment by ezyang** [comment]
@DavidNemeskey Oops, I missed your message and kept editing the comment above. There are three more questions, if you don't mind ;)

Re 2: Well, in the experiments I ran there, the weights are being initialized randomly, differently, which easily explains the difference in perplexity. But this is *far far* more difference than what you reported in your experiments. And when the initial loss is this far out, I would guess that the weight initialization is not being done correctly. I can't easily tell, because I can't tell if TF is actually successfully loading the saved weights, because the TF script is nondeterministic.

I also ran one more experiment, which is I took your new loss function and slapped it on `word_language_model`, using the wikitext data set. The model trained correctly but the perplexity is too large, probably because the formula is not correct. But I am concerned because this doesn't match your experiments at all.

**Comment by DavidNemeskey** [comment]
@ezyang The formula is correct -- it returns the same loss as TF (at least for the first few iterations...). The difference between it and the loss used in `word_language_model` is that normally, BPTT loss is summed along time steps (like in TF / my `SequenceLoss`), while the latter averages along both axes. So basically, `SequenceLoss == CrossEntropyLoss * num_steps`. I checked. :)

As for not being able to see if the parameters were loaded correctly -- just run the scripts with `-T 1`...

**Comment by ezyang** [comment]
OK, looking at `-T 1` logs it looks like the decoder weights (PyTorch) / softmax weights (TensorFlow) diverge immediately on the first iteration. At least, I'm pretty sure that's the correspondence (that's the `einsum` in TensorFlow.) Do you see that as well?

**Comment by DavidNemeskey** [comment]
@ezyang Thanks for catching the error in the gradient clipping code in the Pytorch version -- turns out that was the culprit. Seems like I should have gone to the forums first...

On the other hand, this is very good news, because I can switch to Pytorch without concerns.
